paper-server/patches/sources/net/minecraft/server/network/ServerGamePacketListenerImpl.java.patch

```diff
@@ -1478,101 +1476,77 @@
-+                if (this.player.hasDisconnected()) {
-+                    return;
-+                }
-+                // CraftBukkit end
++                if (this.player.hasDisconnected()) return; // CraftBukkit - SPIGOT-7346: Prevent disconnected players from executing commands
                  this.performSignedChatCommand(packet, optional.get());
 -                this.detectRateSpam();
--            });
 +                this.detectRateSpam("/" + packet.command()); // Spigot
-+            }, true); // CraftBukkit - sync commands
+             });
          }
      }
  
      private void performSignedChatCommand(ServerboundChatCommandSignedPacket packet, LastSeenMessages lastSeenMessages) {
-+        // CraftBukkit start
-+        String command = "/" + packet.command();
-+        if (org.spigotmc.SpigotConfig.logCommands) { // Paper - Add missing SpigotConfig logCommands check
-+            LOGGER.info("{} issued server command: {}", this.player.getScoreboardName(), command);
-+        } // Paper - Add missing SpigotConfig logCommands check
++        // Paper start
++        final String slashCommand = "/" + packet.command();
++        if (org.spigotmc.SpigotConfig.logCommands) {
++            LOGGER.info("{} issued server command: {}", this.player.getScoreboardName(), slashCommand);
++        }
 +
-+        PlayerCommandPreprocessEvent event = new PlayerCommandPreprocessEvent(this.getCraftPlayer(), command, new org.bukkit.craftbukkit.util.LazyPlayerSet(this.server));
-+        this.cserver.getPluginManager().callEvent(event);
-+        command = event.getMessage().substring(1);
++        final org.bukkit.event.player.PlayerCommandPreprocessEvent event = new org.bukkit.event.player.PlayerCommandPreprocessEvent(this.getCraftPlayer(), slashCommand, new org.bukkit.craftbukkit.util.LazyPlayerSet(this.server));
++        event.callEvent();
++        // Paper end
 +
          ParseResults<CommandSourceStack> parseResults = this.parseCommand(packet.command());
  
          Map<String, PlayerChatMessage> map;
-         try {
-+            // Paper - Always parse the original command to add to the chat chain
-             map = this.collectSignedArguments(packet, SignableCommand.of(parseResults), lastSeenMessages);
-         } catch (SignedMessageChain.DecodeException var6) {
-             this.handleMessageDecodeFailure(var6);
+@@ -1433,11 +_,25 @@
              return;
          }
  
-+        // Paper start - Fix cancellation and message changing
-+        if (event.isCancelled()) {
-+            // Only now are we actually good to return
++        // Paper start
++        // Parsing commands to be on main :cryingandshaking: - if that weren't the case, both this and chat could be pushed off to the chat message chain immediately
++        // Always decode message arguments before cancelling
++        if (event.isCancelled() || this.player.isRemoved()) {
```

Comment:
If we are checking for player removal, I do not think that we should be calling it this far down? 
Cause that event is being fired still.